horovod/mxnet/mpi_ops.cc

```diff
@@ -185,24 +192,49 @@ void DoHorovodOperationCudaOnCPU(void*, void* on_complete_ptr, void* param) {
 inline void PushHorovodOperationCudaOnCPU(OperationType op_type, NDArray* input,
                                           NDArray* output, const char* name,
                                           int priority, int root_rank = -1) {
+                                            
+  LOG(TRACE)<<"PushHorovodOperationCudaOnCPU";
   auto op_type_name = GetOpTypeName(op_type);
   auto op_name = GetOpName(op_type_name, name);
   auto hvd_cpu_buffer = std::make_shared<MXTemporaryBuffer<NDArray>>(
       CPU_DEVICE_ID, input->dtype());
-  auto ops_param = CreateMpiOpsParam(nullptr, nullptr, hvd_cpu_buffer,
+  auto hvd_output_buffer = std::make_shared<MXTemporaryBuffer<NDArray>>(
+      CPU_DEVICE_ID, input->dtype());
+  auto ops_param = CreateMpiOpsParam(nullptr, nullptr, hvd_cpu_buffer, hvd_output_buffer, 
                                      op_type, op_name, root_rank);
 
   // Make async copy of input tensor to CPU tensor.
   TensorUtil::AsyncCopyCudaToCPU(input, hvd_cpu_buffer->tensor());
 
   // In-place
   auto cpu_tensor_var = hvd_cpu_buffer->tensor()->var();
+  auto output_tensor_var = hvd_output_buffer->tensor()->var();
+  auto input_var = input->var();
+  auto output_var = output->var();
+
+  std::vector<Engine::VarHandle> const_vars;
+  std::vector<Engine::VarHandle> mutable_vars;
+
+  const_vars.push_back(input_var);
+  const_vars.push_back(cpu_tensor_var);
+
+  mutable_vars.push_back(output_tensor_var);
+  mutable_vars.push_back(output_var);
+
   MXEnginePushAsync(DoHorovodOperationCudaOnCPU, ops_param, DeleteMpiOpsParam,
-                    &MX_EXEC_CTX, nullptr, 0, &cpu_tensor_var, 1,
-                    &MX_FUNC_PROP, priority, op_type_name);
+                    &MX_EXEC_CTX, const_vars.data(), 2, mutable_vars.data(), 2,
+                    &MX_FUNC_PROP, priority, op_type_name, true);
 
   // Make async copy of CPU tensor to output tensor.
-  TensorUtil::AsyncCopyCPUToCuda(hvd_cpu_buffer->tensor(), output);
+  if (op_type == OperationType::ALLGATHER){
+    hvd_output_buffer->tensor()->WaitToRead();
+    // Preventing exiting the API call prematurely and causing shape issue.
+    TensorUtil::AsyncCopyCPUToCuda(hvd_output_buffer->tensor(), output);
+  }else{
+    hvd_cpu_buffer->tensor()->WaitToRead();
```

Comment:
The async actions are handled by MXNet engine. No need to synchronize here.

Reply:
We may need to add a blocking wait for the Horovod operation in the pushed MXNet operator to remove this wait.